### PR TITLE
Remove lone identity gates in Optimize 1q decomposition pass

### LINKED ISSUE #5287: Qiskit transpile different basis gate set
### Informations
- **Qiskit version**:  
{'qiskit-terra': '0.16.0', 
'qiskit-aer': '0.7.0', 
'qiskit-ignis': '0.5.0',
 'qiskit-ibmq-provider': '0.11.0',
 'qiskit-aqua': '0.8.0', 
'qiskit': '0.23.0'}
- **Python version**: Python 3.8.1
- **Operating system**: macOs

### What is the current behavior?

Currently, the output from the qiskit transpile function for my random gate set (random gates) with the basis gate set 'rx', 'ry' and 'cx' is this circuit: 

Bildschirmfoto 2020-10-25 um 17 56 11

I'm not sure, why the transpile function gives sometimes three single qubit rotations (Is it not possible to decompose every  single qubit rotation into one  rx( -pi <= theta<= pi)  and one ry( -pi <= theta <= pi) rotation ?).

The second minor issue is that sometimes a rotation with an angle of 0 degree is still in the circuit and not removed (see image).  

### Steps to reproduce the problem

The following code produces the circuit of the provided image. 
`
qc = QuantumCircuit(3)

qc.h(0)
qc.h(1)
qc.x(0)
qc.h(0)
qc.ry(np.pi/4, 0)
qc.rx(np.pi/10, 0)
qc.y(1)
qc.rx(np.pi/10, 0)
qc.rx(np.pi/10, 0)
qc.h(2)
qc.cx(2, 1)
qc.x(0)
qc.cx(0, 1)
qc.y(2)
qc.rx(np.pi/10, 0)
qc.rx(np.pi/10, 0)
qc.y(0)
qc.x(2)
qc.rx(np.pi/10, 0)
qc.ry(np.pi/4, 2)
qc.cx(0, 2)
qc.h(2)
qc.cx(2, 1)
qc.x(0)
qc.cx(0, 1)
qc.rx(np.pi/10, 0)
qc.rx(np.pi/10, 0)

qc.x(2)
qc.ry(np.pi/4, 2)
qc.ry(np.pi/4, 2)
qc.ry(np.pi/4, 2)
qc.cx(0, 2)
qc.h(2)
qc.cx(2, 1)
qc.measure_all()

coupling_string = [[0, 1], [1, 0], [2, 0], [0,2], [1,2], [2, 1]]
CM = CouplingMap(coupling_string)
basis_gates = ['id', 'ry', 'rx', 'cx']
transpiled_qc = transpile(qc, coupling_map=CM, basis_gates=basis_gates, optimization_level=3, seed_transpiler=1)

### What is the expected behavior?

Normally I would expect that the transpile function with the passed basis gate set would give me in my circuit only two consecutive applied single qubit roations and not three and that rotations with an angle of 0 would be removed.  

Am I wrong with my expectations ?

### Suggested solutions
...

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

When running the optimize 1q decomposition pass for basis sets other
than u1, u2, u3 there was an edge case where it wouldn't remove lone
identity gates. This was because normally the pass skips runs of length
1 because the Euler decomposer will normally not do better than that.
But, in the case of identity gates we know that they can just be
removed, and the simplify=True flag on the decomposer will normally do
this for us. But, since we don't run the decomposer this commit fixes
this edge case by explicitly checking the single gate for an identity
and removing it if it is.

### Details and comments

Fixes #5287

### CODE REVIEW

File: test/python/transpiler/test_optimize_1q_decomposition.py [hunk 1]
```diff
@@ -223,6 +223,86 @@ def test_parameterized_expressions_in_circuits(self, basis):
...
+        passmanager = PassManager()
+        passmanager.append(BasisTranslator(sel, basis))
+        passmanager.append(Optimize1qGatesDecomposition(basis))
+        result = passmanager.run(circuit)
+        self.assertEqual([], result.data)
+
+    def test_identity_u3(self):
+        """Test lone identity gates in u3 basis are removed."""
+        circuit = QuantumCircuit(1)
+        circuit.u3(0, 0, 0, 0)
```
Comment (Cryoris) [inline test/python/transpiler/test_optimize_1q_decomposition.py:275]: The windows tests seem to be failing due to the deprecation warning emitted for the u3 method, which contains greek ASCII symbols. Not using that method should fix it:
```suggestion
        from qiskit.circuit.library import U3Gate
        circuit = QuantumCircuit(1)
        circuit.append(U3Gate(0, 0, 0), [0])
```

Comment (mtreinish) [reply test/python/transpiler/test_optimize_1q_decomposition.py:275]: Ok, it's a bit odd though that the decoding error is happening on a different process in the test runner, But I'll give this a shot.

Comment (mtreinish) [reply test/python/transpiler/test_optimize_1q_decomposition.py]: Done in: https://github.com/Qiskit/qiskit-terra/pull/5292/commits/13633d8568ff0bb98ccf9d41920764655c488380

File: qiskit/transpiler/passes/optimization/optimize_1q_decomposition.py [hunk 2]
```diff
@@ -70,8 +70,13 @@ def run(self, dag):
         runs = dag.collect_runs(self.euler_basis_names[self.basis])
         runs = _split_runs_on_parameters(runs)
         for run in runs:
-            # Don't try to optimize a single 1q gate
             if len(run) <= 1:
+                params = run[0].op.params
+                # Remove single identity gates
+                if run[0].op.name in self.euler_basis_names[self.basis] and len(
+                        params) > 0 and all(param == 0 for param in params):
```
Comment (kdk) [inline qiskit/transpiler/passes/optimization/optimize_1q_decomposition.py:77]: The `all(param == 0 ...)` check may miss some cases (e.g. for R gates). Can we check something like `np.array_equal(op.to_matrix(), np.eye(2))` instead?

Comment (mtreinish) [reply qiskit/transpiler/passes/optimization/optimize_1q_decomposition.py]: Yeah, that's a good idea. Done in: https://github.com/Qiskit/qiskit-terra/pull/5292/commits/0fdfcce433ac1a800938344e0fb6c7a3335cb7d3

**Review by kdk — APPROVED** [review]

**Pull request merged**